[MTS/TC] [70405604] Add change handler

lfender6445 opened the pull request:
Pivotal Story 1: https://www.pivotaltracker.com/story/show/70405604
Pivotal Story 2: https://www.pivotaltracker.com/story/show/76127754

We're introducing some functionality that will allow us to tag change events on `<select>` elements. We extracted some functionality from the base object in this library. Now, there are two CoffeeScript classes; one deals with clicks, and the other deals with change events. Using RequireJS in this library to include those new CoffeeScript files means we will have to edit the RequireJS configuration files of our applications. We'll have to  include jquery-autotagging in the `packages` section instead of the `paths` section.
- [Merge Me First](https://github.com/primedia/ag/pull/1663)
- We MUST ensure we update `bower.json` and the RequireJS configuration files for:
  - ag
  - ag_sites
  - m.apartmentguide
  - m.newhomeguide
  - ag_mobile_api
  - wp_blog_themes
  - rent

All sites that use jquery-autotagging need to be updated. Rentals uses webtrends so we can ignore it.

michaelstalker committed:
[MTS] [70476256] Refactor subgroup id code

Make code a bit cleaner. Letting jQuery filter by using #closest()
is preferable to us iterating through a collection. The new way
may be faster, too.

michaelstalker committed:
[MTS] [70405604] Refactor specs

I removed a duplicate spec. I also moved some method names into
"describe" blocks for consistency and ease of identification.

michaelstalker committed:
[MTS] [70405604] Refactor #getItemId

I think my one line is clearer than the four that were there.

- [x] Jasmine specs pass

michaelstalker committed:
[MTS] [70405604] Move functionality to new object

The WH object was already responsible for more than it
should have been. In order to make it easier to add a
change event handler, I created a click event handler.

michaelstalker committed:
[MTS] [70405604] Move click handler into module

michaelstalker committed:
[MTS] [70405604] Remove extraneous code

michaelstalker committed:
[MTS] [70405604] Create 'bind' public interface

I altered the code so we can now add event handlers to
the autotagging library easily. We only have to add a
new event handler type to the #eventHandlers() method.
All event handlers should implement the #bind() method.

michaelstalker committed:
[MTS] [70405604] Create handler for change events

This handler will just handle change events made by "select"
elements. We don't need to handle change events made by inputs
or textareas for now.

michaelstalker committed:
[MTS/CW] [76127754] Improve code clarity

Caleb Wright made some suggestions in a PR, and this commit
addresses those.

michaelstalker committed:
[MTS/TC] [76127754] Imrove code clarity again

* Remove unused method
* Swap body in two methods for better readability. One method just
  called the other.

michaelstalker committed:
[MTS/TC] [76127754] Fix bug in event handling

The old code triggered a custom success and a custom error
event before the image had finished loading. We don't think
this was the intention of the original code author. We made
callback functions that corrected this.

michaelstalker committed:
[MTS/TC] [76127754] Specify bound element

We can now specify which element the tracking pixel success
and error events get bound to. Before, it had to be bound to
the body.

If we specify an element, and there is no client code to handle
the event, the event will bubble up to the body, so nothing
should break. An exception may be if the client code stops event
propagation, and an ancestor element expects to receive the
event. That scenario seems to be outside the responsibility of
the autotagging library, though.

michaelstalker committed:
[MTS] [76127754] Update README

tommyc committed:
Merge branch 'dev' into mts_70405604_add_change_handler

Conflicts:
	jquery.autotagging.js
	src/jquery.autotagging.coffee

lfender6445 committed:
[LF] [73070230] Document.location -> window.location.host

lfender6445 committed:
[LF] [000000] Bump version

lfender6445 committed:
Revert "[LF] [73070230] Document.location -> window.location.host"

This reverts commit 68c7151d569bef060afa480ae2c35157b302ddd2.

lfender6445 committed:
[LF] [000000] Remove unusued method

michaelstalker commented on bower.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "jquery-autotagging",
-  "version": "1.0.15",
+  "version": "1.0.16",
```
@tommyc: @pasha had suggested bumping the major version. We need to change the RequireJS configuration files in each app that uses this, so our code counts as a breaking change. What do you think?

lfender6445 replied:
Yes sounds like a plan to me!

michaelstalker replied:
@lfender6445 I bumped the version to 2.0.0 in `bower.json` and `package.json`.

michaelstalker commented:
@lfender6445 says: @tommyc @mstalker - don't forget to push a tag post merge. i've already bumped version in bower.json

michaelstalker committed:
[MTS] [76127754] Bump version number

The changes to the library require some changes to client applications,
so I bumped the major version number.

tommyc committed:
[TC][70405604] fix missing parameter

pasha commented:
:+1:
